### ISSUE TITLE
Congestions in Munich

### ISSUE THREAD

**Issue by jankaeh** [openego/eTraGo#153]
Running pf and lopf calculations with pypsa I have to cope with congestions/not converging calculations.
Because Munich somehow always seemed to make the biggest problems, I took a look at the overview map provided by Stadtwerke München (https://www.swm-infrastruktur.de/dam/swm-infrastruktur/dokumente/strom/netzstrukturdaten/strukturmerkmale/swm-infrastruktur-schema.pdf) and compared it to the grid model. It seems that there is missing a 380/110-kV-transformer at Heizkraftwerk Nord (bus0="23648", bus1="16573"). It looks like the 110-kV-bus is situated outside the way of the substation:
![muenchen_heizkraftwerk_nord_fehlender_trafo](https://user-images.githubusercontent.com/33579567/39717468-a1fb4b02-5233-11e8-8ba1-0e77dda64fae.png)
Further more it looks like many lines are missing:
![muenchen_ausschnitt](https://user-images.githubusercontent.com/33579567/39717753-7db35824-5234-11e8-9e62-d68f5a2913de.jpg)

I wrote a function to add a trafo and the lines I could identify without doubt in the linked overview map.
In addition to lopf with extendable lines, this helped to successfully do a network.pf() for snapshots that were not converging before.

If you are interested I can upload it. Did you try to apply similar post processing methods? I see in http://iopscience.iop.org/article/10.1088/1742-6596/977/1/012003/pdf that you experienced similar problems.

**Comment by ulfmueller** [comment]
Nice work @jankaeh. Yes, we experience(d) these pf (and lopf) feasibility problems as well. We used a load shedding approach to identify the 'locations' of the infeasibilities (concerning the lopf). As well our conclusions were that it happens in areas with tense populations such as big cities (Hamburg, Munich, Berlin,...). The reason seems to be the high amount of underground cables of which the data base in osm is rather bad. To solve these problems we developed a bit more generic/general approach in which we set the all lines and transformers extendable in a lopf giving them very high capital costs in order to not compete with the dispatch. As a result we get post-processed maximal line capacities. 
Cool, that 'only' by fixing the problems in Munich you were able to solve the infeasibilities. Is your code already in your fork, so I could have a little look at it? 
I think I will discuss this issue with some colleagues and then we'll see if we also use and appreciate your manual work or if we stick to the other solution (or if we combine them).

**Comment by ulfmueller** [comment]
@BartelsJ @lukasol @MarlonSchlemminger If you have any opinions on this - let me know. See also the PR #154.

**Comment by jankaeh** [comment]
Note that I ran a lopf with line extensions before, but still had problems with not converging pf. That's why I took a closer look at Munich:
![2011-01-08 09 00 00](https://user-images.githubusercontent.com/33579567/39759626-7704ad26-52d3-11e8-8e5d-70d9e4e4e2ce.png)
I wanted to do the same for Stuttgart, but as the 110-kV-lines in the city are owned by a TSO not by a DSO there is no map availabe. Maybe the obligation to publish a map of the hv-grid does not count for DSOs explicitly.

Are there plans to make post-processed lines available in the grid schema of the oep? I only ran a lopf for a few days, so my line extensions probably are poor. For example there were no congestions for Berlin.

**Comment by jankaeh** [comment]
* in the plot bus0 and bus1 of extended lines for that snapshot are blue

**Comment by ulfmueller** [comment]
good to know.

> Are there plans to make post-processed lines available in the grid schema of the oep? I only ran a lopf for a few days, so my line extensions probably are poor. For example there were no congestions for Berlin.

yes we are in the process of running an lopf for every single hour for the Status Quo scenario. But the plan was not to introduce the changes in the OEP but rather provide these corrected s_nom as a csv in the eTraGo repos which you can read to overwrite the s_nom values. The complete run will take about 3-4 days. We will give notice when it ran successfully.

**Comment by jankaeh** [comment]
I found a map for Stuttgart and environment, which I used together with Qgis to identify several problems (although this map is much worse than the one from Munich: https://assets.ctfassets.net/xytfb1vrn7of/NZO8x4rKesAcYGGcG4SQg/b780d6a3ca4c2600ab51a30b70950bb1/netzschemaplan-110-kv.pdf).
I added missing connections. It seems that in this particular case there are some 110-kV-overhead-lines mapped but not included in the grid topology. I guess because the last meters are underground cables.
(And taking a look at the topology around Kassel, I found a missing connection between 220-kV-lines.)

The transformer added in the function data_manipulation_sh(network) has a huge x. I don't know where the x values come from, but normally it is 0.135/s_nom. I changed this in data_manipulation_sh() and use this calculation in my function now too. As the function is not specific for Munich anymore I changed the name.

On my system, I now first apply this function to manually add transformers and lines and then run a lopf with line extension. The amount of necessary line extension in Munich and Stuttgart is reduced, although those areas are still critical:
![2011-01-08 09 00 00](https://user-images.githubusercontent.com/33579567/40057832-ac345b28-584f-11e8-9003-2152350a9fee.png)
